### DISCUSSION TITLE
Modifications

### DISCUSSION THREAD

**Discussion by pmatvos** [theyosh/TerrariumPI#680] [Q&A] [answered]
I usually don't code in Python and i'm never coding user apps. So I apologize if some questions will be stupid. 🤦🏻

1. I have day/night light. And to switched it from day to night mode I have to turn it off, on, off and on again. and between on and off there musn't be more than 5s delay. The timer in TP4 can only be set for minutes as far as I know. So I wrote my one function for the night mode and add the correct sequence.   It takes a long time to recompile the whole software. Not a good way for testing. How can I compile core without all the modules or how can I comunicate with api ( to add the function as a module)? So the testing could go faster.

2. I have rpi 3 b+ and gpio relay, I have connected my devices to a relay and set them up in the software. But when i turned of the pi and switched it back on I couldn't connect to web interface any more. But i could still connect to the pi thrue ssh. I know that TP was still running in the background, because al the timers I have set up were still working, but I couldn't connect to web interface. I have tried restarting the servis (sudo service terrariumpi restart),  restarting network interface and pi but nothing work. So i had to reinstall the software. Now it work fine. Is there any other solution for this problem besides reinstalling,  if it happens again? Dose restarting the servis also restarts the web interface? 

3. Is it possible to make the tests with TP4 and the gpio relay in virtual environment? I don't want for my fish tank to be offline while I'm doing the testing unfortunately I don't have 2 sets of hardware.

**Comment by theyosh** [comment, marked as answer]
Hi,

1. Why is the need for less then 5s delay? If you select in the lights area the relays for days and night, they should toggle. The only thing is that it could take up to 30 seconds (max).
![image](https://user-images.githubusercontent.com/8080177/160427735-01228809-d7aa-4fa9-a4c9-becb11c4ab61.png)
![image](https://user-images.githubusercontent.com/8080177/160427776-d5d9d70a-01bb-45bb-a430-f56c5dc4a2f7.png)
In the above example, the delay between day and night would be max 30 secs. But it could be less. I have not tested this that precise.
![image](https://user-images.githubusercontent.com/8080177/160428512-d4962c83-5f3e-43f8-bfe4-babd4433aa32.png)
The relays are toggled close to each other. (Timing is not correct, as the relays using weather data)
And there is no compiling involved, as all code is Python. So not sure what you are doing... Do not use Docker for developing, that does not work. Use a manual install for that: https://theyosh.github.io/TerrariumPI/install/#manual

2. Can't tell what is happening without errors. Running it manual should show the errors: https://theyosh.github.io/TerrariumPI/faq/how-to-debug-terrariumpi/ (Again no Docker)

3. No, not possible. Testing should always be done on a separate PI. I do know that it is hard to get one nowadays, as there is a chip shortage.

**Reply by pmatvos** [reply]
Less than 5s delay is needed because the light has 3 setting. White light, white and blue light, and blue light. To switch between them you need to switch it of and on again. But if you wait for more then 5 s it won't switch it will be white light again. ☹️

I'm editing code with Atom. Jes I did run the installer manually but it takes 30 min to finish. 
So every time I want to test the changes I have to run the installer, is there no faster way?

**Reply by theyosh** [reply]
I do not know what you are changing, but adding new python code does not require a reinstall. So you should just be able to add new code, and restart the software.

But, I would suggest to make a single script that can toggle your light. And use that as a script relay. I think you should make multiple scripts, each for each color. And then add 3 script relays and use them as needed/wanted. If the scripts are working when you run them manual on the console, then add them to TP4
